community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanWithTail.scala

```diff
@@ -66,32 +64,51 @@ case class PlanWithTail(expressionRewriterFactory: (LogicalPlanningContext => Re
 
   override def apply(lhs: LogicalPlan, remaining: Option[PlannerQuery])(implicit context: LogicalPlanningContext): LogicalPlan = {
     remaining match {
-      case Some(query) =>
+      case Some(plannerQuery) =>
         val lhsContext = context.recurse(lhs)
-        val partPlan = planPart(query, lhsContext, Some(context.logicalPlanProducer.planQueryArgumentRow(query.queryGraph)))
+        val partPlan = planPart(plannerQuery, lhsContext, Some(context.logicalPlanProducer.planQueryArgumentRow(plannerQuery.queryGraph)))
         ///use eager if configured to do so
         val alwaysEager = context.config.updateStrategy.alwaysEager
         //If reads interfere with writes, make it a RepeatableRead
+        var shouldPlanEagerBeforeTail = false
         val planWithEffects =
-          if (alwaysEager || (query.updateGraph overlaps query.queryGraph))
-            context.logicalPlanProducer.planRepeatableRead(partPlan)
+          if (!plannerQuery.isInstanceOf[MergePlannerQuery] &&
```

Comment:
This part is pretty hard to grasp, could we simplify? Getting rid of `isInstanceOf` would be nice as well.